src/main/scala/beam/agentsim/agents/ridehail/MatchmakingUtils.scala

```diff
@@ -0,0 +1,394 @@
+import beam.agentsim.agents.ridehail.RideHailVehicleManager.RideHailAgentLocation
+import beam.agentsim.agents.vehicles.EnergyEconomyAttributes.Powertrain
+import beam.agentsim.agents.vehicles.{BeamVehicle, BeamVehicleType, PersonIdWithActorRef}
+import beam.agentsim.agents._
+import beam.router.BeamRouter.Location
+import beam.router.BeamSkimmer
+import beam.router.Modes.BeamMode
+import beam.sim.common.GeoUtils
+import beam.sim.config.BeamConfig.Beam.Agentsim.Agents.RideHail.AllocationManager
+import beam.sim.{BeamServices, Geofence}
+import com.vividsolutions.jts.geom.{Coordinate, GeometryFactory}
+import org.geotools.referencing.GeodeticCalculator
+import org.geotools.referencing.crs.DefaultGeographicCRS
+import org.matsim.api.core.v01.population.Activity
+import org.matsim.api.core.v01.{Coord, Id}
+import org.matsim.core.population.PopulationUtils
+
+import scala.collection.immutable.List
+import scala.collection.mutable.ListBuffer
+
+object MatchmakingUtils {
+
+  def computeAlonsoMoraCost(trip: RideHailTrip, vehicle: VehicleAndSchedule): Double = {
+    val empty = vehicle.getSeatingCapacity - trip.requests.size - vehicle.getNoPassengers
+    val largeConstant = trip.upperBoundDelays
+    val cost = trip.sumOfDelays + empty * largeConstant
+    cost
+  }
+
+  def computeGreedyCost(trip: RideHailTrip, vehicle: VehicleAndSchedule): Double = {
+    val passengers = trip.requests.size + vehicle.getNoPassengers
+    val delay = trip.sumOfDelays
+    val maximum_delay = trip.upperBoundDelays
+    val cost = passengers + (1 - delay / maximum_delay.toDouble)
+    -1 * cost
+  }
+
+  def checkAngle(origin: Coord, dest1: Coord, dest2: Coord)(implicit services: BeamServices): Boolean = {
+    val crs = DefaultGeographicCRS.WGS84
+    //val crs = MGC.getCRS(services.beamConfig.beam.spatial.localCRS)
+    val orgWgs = services.geo.utm2Wgs.transform(origin)
+    val dst1Wgs = services.geo.utm2Wgs.transform(dest1)
+    val dst2Wgs = services.geo.utm2Wgs.transform(dest2)
+    //val crs = CRS.decode(services.beamConfig.beam.spatial.localCRS)
+    val calc = new GeodeticCalculator(crs)
+    val gf = new GeometryFactory()
+    val point1 = gf.createPoint(new Coordinate(orgWgs.getX, orgWgs.getY))
+    calc.setStartingGeographicPoint(point1.getX, point1.getY)
+    val point2 = gf.createPoint(new Coordinate(dst1Wgs.getX, dst1Wgs.getY))
+    calc.setDestinationGeographicPoint(point2.getX, point2.getY)
+    val azimuth1 = calc.getAzimuth
+    val point3 = gf.createPoint(new Coordinate(dst2Wgs.getX, dst2Wgs.getY))
+    calc.setDestinationGeographicPoint(point3.getX, point3.getY)
+    val azimuth2 = calc.getAzimuth
+    val degrees = azimuth2 - azimuth1
+    Math.abs(degrees) < 45.0
+  }
+
+  def checkDistance(r: MobilityRequest, schedule: List[MobilityRequest], searchRadius: Double): Boolean = {
+    schedule.foreach { s =>
+      if (GeoUtils.distFormula(r.activity.getCoord, s.activity.getCoord) <= searchRadius)
+        return true
```

Comment:
returns in scala are *very* much a no no. Not just for syntax but functionality can be different than you expect. Here I would use `exists`:

```
schedule.exists(s => GeoUtils.distFormula(r.activity.getCoord, s.activity.getCoord) <= searchRadius)
````

That should be the only thing you need.